tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,424 @@
+func (w *ReOrgWatcher) watchTransactions() {
+	defer w.Wg.Done()
+
+	runCtx, cancel := w.WithCtxQuitNoTimeout()
+	defer cancel()
+
+	newBlockChan, blockErr, err := w.cfg.ChainBridge.RegisterBlockEpochNtfn(
+		runCtx,
+	)
+	if err != nil {
+		w.reportErr(fmt.Errorf("unable to register for block "+
+			"epoch notifications: %w", err))
+		return
+	}
+
+	for {
+		select {
+		case newProofs := <-w.incomingProofs:
+			txHash := newProofs.proofs[0].AnchorTx.TxHash()
+
+			log.Infof("Watching new proof anchor TX %v for %d "+
+				"assets until it reaches %d confirmations",
+				txHash, len(newProofs.proofs), w.cfg.SafeDepth)
+
+			err := w.waitForConf(runCtx, newProofs)
+			if err != nil {
+				w.reportErr(err)
+				return
+			}
+
+		case conf := <-w.incomingConfs:
+			txHash := conf.Tx.TxHash()
+			txNtfn, ok := w.pendingProofs[txHash]
+			if !ok {
+				w.reportErr(fmt.Errorf("received confirmation "+
+					"for unknown anchor TX %v", txHash))
+				return
+			}
+
+			// If the transaction was confirmed normally, we don't
+			// need to do anything and can stop watching it.
+			log.Infof("Anchor TX %v was confirmed at height %d "+
+				"(block_hash=%v), updating %d proof(s)", txHash,
+				conf.BlockHeight, conf.BlockHash,
+				txNtfn.numProofs())
+
+			// We should never accept an empty proof slice to watch,
+			// but we add this check just to prevent panics in case
+			// we use the watcher incorrectly somewhere.
+			if txNtfn.numProofs() == 0 {
+				w.reportErr(fmt.Errorf("received confirmation "+
+					"for anchor TX %v with empty proof "+
+					"slice", txHash))
+				return
+			}
+
+			// Let's make sure we got a new block hash. If not,
+			// something is weird and would indicate lnd is
+			// misbehaving.
+			if conf.BlockHash == nil {
+				w.reportErr(fmt.Errorf("received confirmation "+
+					"for anchor TX %v with nil block hash",
+					txHash))
+				return
+			}
+			confHash := *conf.BlockHash
+
+			// Do we actually have a different block than the one
+			// we already have in the proof? If not, we can just
+			// ignore this confirmation (this should not happen in
+			// normal circumstances and would be an indication of
+			// the chain backend being misconfigured).
+			firstProof := txNtfn.firstProof()
+			if firstProof.BlockHeader.BlockHash() == confHash {
+				log.Warnf("Anchor TX %v was already confirmed "+
+					"in block %v, ignoring confirmation "+
+					"for block %v", txHash,
+					conf.BlockHeight, conf.BlockHash)
+				continue
+			}
+
+			// We can now update the proofs with the new block and
+			// inform the caller if necessary.
+			err = w.updateProofs(txNtfn, conf)
+			if err != nil {
+				w.reportErr(fmt.Errorf("error updating "+
+					"proofs: %w", err))
+				return
+			}
+
+		case newBlock := <-newBlockChan:
+			log.Infof("New block at height %d", newBlock)
+			w.bestHeight = uint32(newBlock)
+
+			for txid := range w.pendingProofs {
+				proofNtfn := w.pendingProofs[txid]
```

Comment:
nit: I would move this line down so that it's just before `proofNtfn.cancel()`. We only need `proofNtfn` within the `if` clause (if we're cancelling).

Reply:
No, we also need `proofNtfn` to call `firstRegistration()` to get the block height.